### PR TITLE
Emily V - Leaves

### PR DESCRIPTION
**Pull request by emilyvomacka**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Work.spotlight selects a random work to feature from the pool of works that have received votes. When no votes are present, it selects a random work. When no works are present, it returns nil, an outcome handled in a friendly way by the view. 
Describe how you approached testing that model method. What edge cases did you come up with? | I tested that it returned a work with votes when votes were present in the database. I also tested its handling of the following edge cases: no votes registered, and no works loaded. 
What are `session` and `flash`? What is the difference between them? | The session hash and flash hash both retain limited information pertaining to the user's session with the website. The main difference is persistence: session is retained throughout the browser session, whereas flash only persists for the current (flash.now) or one additional request/response cycle.
What was one thing that you gained more clarity on through this assignment? | Bootstrap and the idea of CSS frameworks in general.
What is the Heroku URL of your deployed application? | http://emily-v-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | A classwide lecture about Bootstrap grid capabilities would be awesome! We missed out on that category during our jigsaw and it seems like very important functionality.

### CODE REVIEW

File: app/views/works/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  <% else %>
+    <section class="media-table">
+    <h4>
+      <%= media.capitalize + "s" %>
+    </h4>
+    <table class="table">
+      <thead>
+        <tr>
+          <th width=80 scope="col">Votes</th>
+          <th scope="col">Title</th>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:17]: Nice! You could also achieve this with form parital

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      render :edit
+      return
+    end 
+  end 
+  
+  def upvote 
+    if session[:user_id].nil? 
+      flash[:warning] = "A problem occurred: You must log in to do that"
+      redirect_to work_path(@work)
+      return
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:46]: This kind of thing is better done with a controller filter.

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | yes, but there was room for more, see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no, see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good work overall! I see that the learning goals for this assignment have been met! I've left some cheers and jeers in your comments for future reference!
